File: src/api/benchmark.ts
import { invokeWithResponseSchema } from "@/api/ipc";
import {
	BenchmarkDraftIdsSchema,
	BenchmarkAssetPreviewSchema,
	BenchmarkSummariesSchema,
	BenchmarkDetailSchema,
	BenchmarkDraftSchema,
	BenchmarkMountSchema,
	BenchmarkMountsSchema,
	BenchmarkFileSchema,
	BenchmarkImportPreviewSchema,
	BenchmarkTagSchema,
	BenchmarkTagsSchema,
	BenchmarkPreviewSchema,
	BenchmarkTaskDetailSchema,
	BenchmarkArtifactFilesSchema,
	BenchmarkArtifactPreviewSchema,
	EmptyBenchmarkResponseSchema,
} from "@/types/benchmark";
import type {
	BenchmarkDocument,
	BenchmarkFilters,
	BenchmarkPreviewInput,
	RerunBenchmarkTaskInput,
	StartBenchmarkTaskInput,
} from "@/types/benchmark";

/**
 * Keeps sorting and filtering ahead of native pagination.
 *
 * @example
 * listBenchmarks(filters, 0);
 */
const listBenchmarks = (filters: BenchmarkFilters, page: number) =>
	invokeWithResponseSchema("list_benchmarks", BenchmarkSummariesSchema, {
		request: { ...filters, page },
	});

/**
 * Loads the pinned version when supplied.
 *
 * @example
 * getBenchmark("suite", "v1");
 */
const getBenchmark = (benchmarkId: string, versionId: string | null = null) =>
	invokeWithResponseSchema("get_benchmark", BenchmarkDetailSchema, {
		request: { benchmarkId, versionId },
	});

/** Hides one personal definition from the default catalog without deleting history.
 * @example archiveBenchmark("suite")
 */
const archiveBenchmark = (benchmarkId: string) =>
	invokeWithResponseSchema("archive_benchmark", BenchmarkDetailSchema, {
		request: { benchmarkId },
	});

/** Loads available classifications. */
const listBenchmarkTags = () =>
	invokeWithResponseSchema("list_benchmark_tags", BenchmarkTagsSchema);

/**
 * Stores an allowed Gravity selection.
 *
 * @example
 * createBenchmarkTag("Coding", "Code");
 */
const createBenchmarkTag = (name: string, icon: string) =>
	invokeWithResponseSchema("create_benchmark_tag", BenchmarkTagSchema, {
		request: { name, icon },
	});

/** Inspects one picker-selected Theoria folder without creating persistence.
 * @example previewBenchmarkImport("/templates/example")
 */
const previewBenchmarkImport = (sourcePath: string) =>
	invokeWithResponseSchema(
		"preview_benchmark_import",
		BenchmarkImportPreviewSchema,
		{ request: { sourcePath } },
	);

/** Copies a reviewed template folder into one editable personal draft.
 * @example importBenchmarkFolder("/templates/example", "tag-1")
 */
const importBenchmarkFolder = (sourcePath: string, tagId: string) =>
	invokeWithResponseSchema("import_benchmark_folder", BenchmarkDraftSchema, {
		request: { sourcePath, tagId },
	});

/** Copies one picker-selected file into managed Benchmark storage.
 * @example importBenchmarkAsset("/tmp/input.txt", "input.txt")
 */
const importBenchmarkAsset = (sourcePath: string, path: string) =>
	invokeWithResponseSchema("import_benchmark_asset", BenchmarkFileSchema, {
		request: { sourcePath, path },
	});

/** Returns a bounded managed-file preview without exposing its native path.
 * @example previewBenchmarkAsset("asset-1")
 */
const previewBenchmarkAsset = (assetId: string) =>
	invokeWithResponseSchema(
		"preview_benchmark_asset",
		BenchmarkAssetPreviewSchema,
		{ request: { assetId } },
	);

/** Opens an opaque managed document with the application selected by the user.
 * @example openBenchmarkAsset("asset-1", "/Applications/Preview.app")
 */
const openBenchmarkAsset = (assetId: string, applicationPath: string) =>
	invokeWithResponseSchema(
		"open_benchmark_asset",
		EmptyBenchmarkResponseSchema,
		{ request: { assetId, applicationPath } },
	);

/** Stores an editor buffer as a new immutable managed file revision.
 * @example saveBenchmarkTextAsset("input.txt", "content")
 */
const saveBenchmarkTextAsset = (path: string, text: string) =>
	invokeWithResponseSchema("save_benchmark_text_asset", BenchmarkFileSchema, {
		request: { path, text },
	});

/**
 * Saves with optimistic concurrency.
 *
 * @example
 * saveBenchmarkDraft(document, "draft", 1);
 */
const saveBenchmarkDraft = (
	document: BenchmarkDocument,
	draftId: string | null,
	expectedRevision: number | null,
	benchmarkId: string | null = null,
) =>
	invokeWithResponseSchema("save_benchmark_draft", BenchmarkDraftSchema, {
		request: {
			document,
			draftId,
			expectedRevision,
			...(benchmarkId ? { benchmarkId } : {}),
		},
	});

/**
 * Publishes the reviewed saved revision.
 *
 * @example
 * publishBenchmark("draft", 2);
 */
const publishBenchmark = (draftId: string, expectedRevision: number) =>
	invokeWithResponseSchema("publish_benchmark", BenchmarkDetailSchema, {
		request: { draftId, expectedRevision },
	});

/**
 * Restores an unfinished editor.
 *
 * @example
 * getBenchmarkDraft("draft");
 */
const getBenchmarkDraft = (draftId: string) =>
	invokeWithResponseSchema("get_benchmark_draft", BenchmarkDraftSchema, {
		request: { draftId },
	});

/**
 * Lists saved editor IDs without loading every document.
 *
 * @example
 * listBenchmarkDrafts(0);
 */
const listBenchmarkDrafts = (page: number) =>
	invokeWithResponseSchema("list_benchmark_drafts", BenchmarkDraftIdsSchema, {
		request: { page },
	});

/**
 * Reads fixed workspace relationships.
 *
 * @example
 * listWorkspaceBenchmarks("workspace", 0);
 */
const listWorkspaceBenchmarks = (workspaceId: string, page: number) =>
	invokeWithResponseSchema("list_workspace_benchmarks", BenchmarkMountsSchema, {
		request: { workspaceId, page },
	});

/**
 * Never silently upgrades an existing mount.
 *
 * @example
 * mountBenchmark("workspace", "suite", "v1");
 */
const mountBenchmark = (
	workspaceId: string,
	benchmarkId: string,
	versionId: string,
) =>
	invokeWithResponseSchema("mount_benchmark", BenchmarkMountSchema, {
		request: { workspaceId, benchmarkId, versionId },
	});

/** Explicitly changes the version pinned by an existing workspace mount.
 * @example updateBenchmarkMount("workspace", "mount", "version-2")
 */
const updateBenchmarkMount = (
	workspaceId: string,
	mountId: string,
	versionId: string,
) =>
	invokeWithResponseSchema("update_benchmark_mount", BenchmarkMountSchema, {
		request: { workspaceId, mountId, versionId },
	});

/** Changes whether one Workspace mount is ordered above ordinary mounts.
 * @example setBenchmarkMountPin("workspace", "mount", true)
 */
const setBenchmarkMountPin = (
	workspaceId: string,
	mountId: string,
	isPinned: boolean,
) =>
	invokeWithResponseSchema("set_benchmark_mount_pin", BenchmarkMountSchema, {
		request: { workspaceId, mountId, isPinned },
	});

/**
 * Removes only the relationship.
 *
 * @example
 * unmountBenchmark("workspace", "mount");
 */
const unmountBenchmark = (workspaceId: string, mountId: string) =>
	invokeWithResponseSchema("unmount_benchmark", EmptyBenchmarkResponseSchema, {
		request: { workspaceId, mountId },
	});

/**
 * Checks the whole suite without model overrides or Task creation.
 *
 * @example
 * previewBenchmarkTask(input);
 */
const previewBenchmarkTask = (request: BenchmarkPreviewInput) =>
	invokeWithResponseSchema("preview_benchmark_task", BenchmarkPreviewSchema, {
		request,
	});

/** Creates one immutable task plan and starts its background execution.
 * @example startBenchmarkTask(input)
 */
const startBenchmarkTask = (request: StartBenchmarkTaskInput) =>
	invokeWithResponseSchema("start_benchmark_task", BenchmarkTaskDetailSchema, {
		request,
	});

/** Lists final files for one execution with changes from its Case baseline.
 * @example listBenchmarkExecutionArtifacts("task-1", "execution-1")
 */
const listBenchmarkExecutionArtifacts = (taskId: string, executionId: string) =>
	invokeWithResponseSchema(
		"list_benchmark_execution_artifacts",
		BenchmarkArtifactFilesSchema,
		{ request: { taskId, executionId } },
	);

/** Reads one bounded final-file preview without exposing its native path.
 * @example previewBenchmarkExecutionArtifact("task-1", "execution-1", "result.txt")
 */
const previewBenchmarkExecutionArtifact = (
	taskId: string,
	executionId: string,
	path: string,
) =>
	invokeWithResponseSchema(
		"preview_benchmark_execution_artifact",
		BenchmarkArtifactPreviewSchema,
		{ request: { taskId, executionId, path } },
	);

/** Creates a fresh Task from one terminal run while preserving its published version.
 * @example rerunBenchmarkTask(input)
 */
const rerunBenchmarkTask = (request: RerunBenchmarkTaskInput) =>
	invokeWithResponseSchema("rerun_benchmark_task", BenchmarkTaskDetailSchema, {
		request,
	});

/** Restores the latest persisted matrix state for one Benchmark Task.
 * @example getBenchmarkTask("task-1")
 */
const getBenchmarkTask = (taskId: string) =>
	invokeWithResponseSchema("get_benchmark_task", BenchmarkTaskDetailSchema, {
		request: { taskId },
	});

export {
	listBenchmarks,
	getBenchmark,
	archiveBenchmark,
	listBenchmarkTags,
	createBenchmarkTag,
	previewBenchmarkImport,
	importBenchmarkFolder,
	importBenchmarkAsset,
	previewBenchmarkAsset,
	openBenchmarkAsset,
	saveBenchmarkTextAsset,
	saveBenchmarkDraft,
	publishBenchmark,
	getBenchmarkDraft,
	listBenchmarkDrafts,
	listWorkspaceBenchmarks,
	mountBenchmark,
	updateBenchmarkMount,
	setBenchmarkMountPin,
	unmountBenchmark,
	previewBenchmarkTask,
	startBenchmarkTask,
	listBenchmarkExecutionArtifacts,
	previewBenchmarkExecutionArtifact,
	rerunBenchmarkTask,
	getBenchmarkTask,
};

File: src-tauri/src/commands/benchmark.rs
use crate::dto::benchmark::{
    BenchmarkDetailResponse, BenchmarkDraftResponse, BenchmarkMountResponse,
    BenchmarkSummaryResponse, BenchmarkTagResponse, ListBenchmarksRequest, MountBenchmarkRequest,
    PublishBenchmarkRequest, SaveBenchmarkDraftRequest,
};
use crate::dto::benchmark::{
    CreateBenchmarkTagRequest, GetBenchmarkDraftRequest, GetBenchmarkRequest,
    ListBenchmarkDraftsRequest, ListWorkspaceBenchmarksRequest, UnmountBenchmarkRequest,
};
use crate::error::IpcError;
use crate::services::benchmark::BenchmarkService;
use tauri::State;

/// Lists local tag names and Gravity icons.
#[tauri::command]
pub(crate) async fn list_benchmark_tags(
    service: State<'_, BenchmarkService>,
) -> Result<Vec<BenchmarkTagResponse>, IpcError> {
    service
        .tags()
        .await
        .map(|items| items.into_iter().map(Into::into).collect())
        .map_err(Into::into)
}
/// Creates a tag for use by personal Benchmark definitions.
#[tauri::command]
pub(crate) async fn create_benchmark_tag(
    request: CreateBenchmarkTagRequest,
    service: State<'_, BenchmarkService>,
) -> Result<BenchmarkTagResponse, IpcError> {
    service
        .create_tag(&request.name, &request.icon)
        .await
        .map(Into::into)
        .map_err(Into::into)
}
/// Persists an incomplete editor document with revision checking.
#[tauri::command]
pub(crate) async fn save_benchmark_draft(
    request: SaveBenchmarkDraftRequest,
    service: State<'_, BenchmarkService>,
) -> Result<BenchmarkDraftResponse, IpcError> {
    service
        .save_draft(
            request.draft_id,
            request.expected_revision,
            request.document,
        )
        .await
        .map(Into::into)
        .map_err(Into::into)
}
/// Loads the selected editor document.
#[tauri::command]
pub(crate) async fn get_benchmark_draft(
    request: GetBenchmarkDraftRequest,
    service: State<'_, BenchmarkService>,
) -> Result<BenchmarkDraftResponse, IpcError> {
    service
        .draft(&request.draft_id)
        .await
        .map(Into::into)
        .map_err(Into::into)
}
/// Returns a bounded page of draft identifiers.
#[tauri::command]
pub(crate) async fn list_benchmark_drafts(
    request: ListBenchmarkDraftsRequest,
    service: State<'_, BenchmarkService>,
) -> Result<Vec<String>, IpcError> {
    service.drafts(request.page).await.map_err(Into::into)
}
/// Publishes the saved and reviewed draft revision.
#[tauri::command]
pub(crate) async fn publish_benchmark(
    request: PublishBenchmarkRequest,
    service: State<'_, BenchmarkService>,
) -> Result<BenchmarkDetailResponse, IpcError> {
    service
        .publish(&request.draft_id, request.expected_revision)
        .await
        .map(Into::into)
        .map_err(Into::into)
}
/// Searches published Benchmark cards.
#[tauri::command]
pub(crate) async fn list_benchmarks(
    request: ListBenchmarksRequest,
    service: State<'_, BenchmarkService>,
) -> Result<Vec<BenchmarkSummaryResponse>, IpcError> {
    service
        .list(
            &request.search,
            request.tag_id.as_deref(),
            request.author.as_deref(),
            request.page,
        )
        .await
        .map(|items| items.into_iter().map(Into::into).collect())
        .map_err(Into::into)
}
/// Shared detail endpoint for catalog and workspace contexts.
#[tauri::command]
pub(crate) async fn get_benchmark(
    request: GetBenchmarkRequest,
    service: State<'_, BenchmarkService>,
) -> Result<BenchmarkDetailResponse, IpcError> {
    service
        .detail(&request.benchmark_id, request.version_id.as_deref())
        .await
        .map(Into::into)
        .map_err(Into::into)
}
/// Pins a published version in the selected workspace.
#[tauri::command]
pub(crate) async fn mount_benchmark(
    request: MountBenchmarkRequest,
    service: State<'_, BenchmarkService>,
) -> Result<BenchmarkMountResponse, IpcError> {
    service
        .mount(
            request.workspace_id,
            request.benchmark_id,
            request.version_id,
        )
        .await
        .map(Into::into)
        .map_err(Into::into)
}
/// Loads a page of workspace mounts.
#[tauri::command]
pub(crate) async fn list_workspace_benchmarks(
    request: ListWorkspaceBenchmarksRequest,
    service: State<'_, BenchmarkService>,
) -> Result<Vec<BenchmarkMountResponse>, IpcError> {
    service
        .mounts(&request.workspace_id, request.page)
        .await
        .map(|items| items.into_iter().map(Into::into).collect())
        .map_err(Into::into)
}
/// Removes a mount without removing content or results.
#[tauri::command]
pub(crate) async fn unmount_benchmark(
    request: UnmountBenchmarkRequest,
    service: State<'_, BenchmarkService>,
) -> Result<(), IpcError> {
    service
        .unmount(&request.workspace_id, &request.mount_id)
        .await
        .map_err(Into::into)
}
